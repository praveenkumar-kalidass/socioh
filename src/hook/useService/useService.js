import * as Keychain from 'react-native-keychain';
import AsyncStorage from '@react-native-community/async-storage';
import RNContacts from 'react-native-contacts';

import { CONSTANT, ERROR, TRANSLATION } from '../../constant';
import useAjax from '../useAjax';
import useUser from '../useUser';

const useService = () => {
  const { ajax } = useAjax();
  const { user, saveUserDetails, clearUserDetails } = useUser();

  const signUp = async ({ name, email, password }) => {
    await Keychain.resetGenericPassword();
    await Keychain.setGenericPassword(email, password);
    await AsyncStorage.setItem(
      CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      JSON.stringify({
        name,
        email,
      }),
    );
    saveUserDetails({
      name,
      email,
    });
    return ajax();
  };

  const signIn = async ({ email, password }) => {
    const credentials = await Keychain.getGenericPassword();
    if (credentials.username === email && credentials.password === password) {
      const userDetails = await AsyncStorage.getItem(
        CONSTANT.ASYNC_STORAGE_KEY.SOCIOH_USER,
      );
      saveUserDetails(JSON.parse(userDetails));
      return ajax();
    }
    throw Error(ERROR.INVALID_CREDENTIALS);
  };

  const logout = async () => {
    clearUserDetails();
    return ajax();
  };

  const getUserDetails = async () => {
    if (user) {
      await ajax();
      return {
        ...user,
        interests: Object.values(TRANSLATION.INTERESTS),
      };
    }
    throw Error(ERROR.NO_USER_FOUND);
  };

  const getContacts = async (search) => {
    const permission = await RNContacts.requestPermission();
    if (permission === CONSTANT.PERMISSION.AUTHORIZED) {
      let contacts;
      if (search) {
        contacts = await RNContacts.getContactsMatchingString(search);
      } else {
        contacts = await RNContacts.getAll();
      }
      await ajax();
      return contacts;
    }
    throw Error(ERROR.UNAUTHORIZED_TO_READ_CONTACTS);
  };

  const getMessages = async () => {
    await ajax();
    return CONSTANT.MESSAGES.GREETINGS;
  };

  const getMessageList = async () => {
    await ajax();
    return CONSTANT.MESSAGES.LIST;
  };

  return {
    signUp,
    signIn,
    logout,
    getUserDetails,
    getContacts,
    getMessages,
    getMessageList,
  };
};

export default useService;
