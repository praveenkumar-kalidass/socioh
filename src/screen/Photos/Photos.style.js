import styled from 'styled-components/native';

import { SPACING } from '../../constant';

export const PhotosContainer = styled.View`
  flex: 1;
  padding-horizontal: ${SPACING.SPACE_20};
`;

export const Gallery = styled.View`
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: ${SPACING.SPACE_20};
`;

export const GalleryGrid = styled.View`
  width: 30%;
  margin-vertical: ${SPACING.SPACE_10};
`;

export const GalleryItem = styled.Image`
  aspect-ratio: 1;
  height: auto;
  width: auto;
`;

export const NoPhotosContainer = styled.View`
  flex: 1;
  justify-content: center;
`;

export const NoPhotos = styled.Text`
  text-align: center;
`;
