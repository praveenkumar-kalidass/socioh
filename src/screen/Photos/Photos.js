import React, { useEffect, useState } from 'react';
import { Card, CardItem } from 'native-base';

import useService from '../../hook/useService';
import { TRANSLATION } from '../../constant';
import {
  Gallery,
  GalleryGrid,
  GalleryItem,
  NoPhotos,
  NoPhotosContainer,
  PhotosContainer,
} from './Photos.style';

const Photos = () => {
  const [images, setImages] = useState([]);

  const { getPhotos } = useService();

  useEffect(() => {
    const loadPhotos = async () => {
      try {
        const photos = await getPhotos();
        setImages(photos);
      } catch (error) {
        console.warn('error', error);
      }
    };
    loadPhotos();
  }, []);

  return (
    <PhotosContainer>
      <Card>
        <CardItem cardBody>
          <Gallery>
            <If condition={!images.length}>
              <NoPhotosContainer>
                <NoPhotos testID="photos_empty">
                  {TRANSLATION.NO_PHOTOS}
                </NoPhotos>
              </NoPhotosContainer>
            </If>
            <For each="image" index="index" of={images}>
              <GalleryGrid key={`photo_${index}`}>
                <GalleryItem
                  testID={`photo_${index}`}
                  source={{ uri: image.uri }}
                />
              </GalleryGrid>
            </For>
          </Gallery>
        </CardItem>
      </Card>
    </PhotosContainer>
  );
};

export default Photos;
