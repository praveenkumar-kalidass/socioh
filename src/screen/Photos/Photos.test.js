import React from 'react';
import { render, waitFor } from '@testing-library/react-native';

import useService from 'hook/useService';
import Photos from './index';

jest.mock('hook/useService');

describe('Photos', () => {
  const mockPhotosService = jest.fn();

  beforeEach(() => {
    mockPhotosService.mockResolvedValue();
    useService.mockImplementation(() => ({ getPhotos: mockPhotosService }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const container = render(<Photos />);

    expect(container).toMatchSnapshot();
  });

  it('should show no photos text when photo list is empty', async () => {
    mockPhotosService.mockResolvedValueOnce([]);

    const { getByTestId } = render(<Photos />);

    await waitFor(() => getByTestId('photos_empty'));

    expect(getByTestId('photos_empty').props.children).toStrictEqual(
      'No Photos available',
    );
  });

  it('should load photos when get photos returns photo list', async () => {
    mockPhotosService.mockResolvedValueOnce([{ uri: 'file://123' }]);

    const { getByTestId } = render(<Photos />);

    await waitFor(() => getByTestId('photo_0'));

    expect(getByTestId('photo_0').props.source).toStrictEqual({
      uri: 'file://123',
    });
  });
});
