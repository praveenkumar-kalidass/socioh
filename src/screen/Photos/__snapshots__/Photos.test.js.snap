// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photos should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
        "paddingHorizontal": 20,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "borderBottomWidth": 0.5,
        "borderColor": "#ccc",
        "borderLeftWidth": 0.5,
        "borderRadius": 2,
        "borderRightWidth": 0.5,
        "borderTopWidth": 0.5,
        "elevation": 3,
        "flexWrap": "nowrap",
        "marginBottom": 5,
        "marginLeft": 2,
        "marginRight": 2,
        "marginTop": 5,
        "shadowColor": "#000",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 1.5,
      }
    }
  >
    <View
      cardBody={true}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#fff",
          "borderRadius": 2,
          "flexDirection": "row",
          "paddingBottom": -5,
          "paddingLeft": -5,
          "paddingRight": -5,
          "paddingTop": -5,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "flexWrap": "wrap",
              "justifyContent": "space-between",
              "paddingBottom": 20,
              "paddingLeft": 20,
              "paddingRight": 20,
              "paddingTop": 20,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "flexBasis": 0,
                "flexGrow": 1,
                "flexShrink": 1,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "textAlign": "center",
                },
              ]
            }
            testID="photos_empty"
          >
            No Photos available
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
